### ISSUE TITLE
[Bug]: Shared links to HTML files are now downloaded as txt files on mobile

### ISSUE THREAD

**Issue by teledyn** [nextcloud/server#49501]
### ⚠️ This issue respects the following points: ⚠️

- [x] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [x] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [x] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [x] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

The best most thorough description I can give, with a screen shot, is a link, so I hope these are allowed: [Mastodon Thread](https://mstdn.ca/@teledyn/113550453173840337)

Since last November, I have been using Nextcloud to share iReal playlists with band members prior to our performances, sometimes during! 

What happens now is the chat flips to the browser where they see a preview of the source code of the html and a 'Download' link. On Linux/Firefox, clicking the link fetches the html file, but on all mobile platforms, the file is _renamed_ as `filename.txt` and not recognized by iReal.

### Steps to reproduce

1. generate a share link for a file with extension `.html` and post to social media
2. click the link and you will get a preview page with a download link
3. select 'Download' and on mobile only you will get a `.txt` file.

for example: [playlist of a few #ireal xmas tunes](https://nas.teledyn.com/s/JtggMFbi5tPYBqT)
(that's on a homecloud, so please don't hammer it)

### Expected behavior

Previously, I would get the share link from my mobile Android NC app, paste it into the group chat, as each member clicked it, they would be offered the iReal app, click it, and iReal would then ask if they wished to add this playlist. It worked beautifully, until this morning. So far as I know, the only change was a recent auto-update to 29.0.6.

### Nextcloud Server version

29

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.3

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Updated from a MINOR version (ex. 28.0.1 to 28.0.2)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [x] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```json
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": {
            "0": "localhost",
            "7": "nextcloudpi",
            "5": "nextcloudpi.local",
            "8": "nextcloudpi.lan",
            "3": "nextcloudpi",
            "11": "173.230.163.232",
            "1": "192.168.0.13",
            "14": "nextcloudpi",
            "2": "nas.teledyn.com",
            "20": "nas.teledyn.com",
            "21": "radio.teledyn.com"
        },
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "29.0.6.1",
        "overwrite.cli.url": "https:\/\/nas.teledyn.com\/",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "tempdirectory": "\/mnt\/ncdata\/data\/tmp",
        "default_phone_region": "CA",
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "preview_max_x": 1024,
        "preview_max_y": 1024,
        "jpeg_quality": 60,
        "overwriteprotocol": "https",
        "maintenance": false,
        "logfile": "\/mnt\/ncdata\/data\/nextcloud.log",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 3,
        "log_type": "file",
        "mail_sendmailmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtptimeout": 20,
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "memories.exiftool": "\/var\/www\/nextcloud\/apps\/memories\/bin-ext\/exiftool-aarch64-glibc",
        "memories.vod.path": "\/var\/www\/nextcloud\/apps\/memories\/bin-ext\/go-vod-aarch64",
        "app_install_overwrite": [
            "podcast",
            "workflow_media_converter",
            "files_opds",
            "news",
            "social",
            "groupfolders",
            "files_archive",
            "data_request",
            "caniupdate",
            "duplicatefinder",
            "files_external_gdrive",
            "issuetemplate",
            "llm",
            "text2image_stablediffusion",
            "metadata",
            "printer"
        ],
        "theme": "",
        "memories.vod.ffmpeg": "\/usr\/bin\/ffmpeg",
        "memories.vod.ffprobe": "\/usr\/bin\/ffprobe",
        "htaccess.RewriteBase": "\/",
        "maintenance_window_start": 9,
        "enable_previews": true,
        "enabledPreviewProviders": {
            "0": "OC\\Preview\\TXT",
            "1": "OC\\Preview\\MarkDown",
            "3": "OC\\Preview\\PDF",
            "6": "OC\\Preview\\Image",
            "7": "OC\\Preview\\Photoshop",
            "8": "OC\\Preview\\TIFF",
            "9": "OC\\Preview\\SVG",
            "10": "OC\\Preview\\Font",
            "11": "OC\\Preview\\MP3",
            "12": "OC\\Preview\\Movie",
            "13": "OC\\Preview\\MKV",
            "14": "OC\\Preview\\MP4",
            "15": "OC\\Preview\\AVI"
        },
        "memories.db.triggers.fcu": true,
        "updater.release.channel": "beta",
        "memories.gis_type": 1,
        "session_keepalive": true,
        "remember_login_cookie_lifetime": 1296000,
        "memories.vod.external": true
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.21.1
  - admin_audit: 1.19.0
  - announcementcenter: 7.0.1
  - app_api: 3.2.1
  - bruteforcesettings: 2.9.0
  - calendar: 4.7.16
  - circles: 29.0.0-dev
  - cloud_federation_api: 1.12.0
  - cloud_py_api: 0.1.9
  - comments: 1.19.0
  - contacts: 6.0.1
  - contactsinteraction: 1.10.0
  - cookbook: 0.11.2
  - dashboard: 7.9.0
  - data_request: 1.16.3
  - dav: 1.30.1
  - federatedfilesharing: 1.19.0
  - federation: 1.19.0
  - files: 2.1.1
  - files_accesscontrol: 1.19.1
  - files_automatedtagging: 1.19.0
  - files_downloadlimit: 2.0.0
  - files_external: 1.21.0
  - files_pdfviewer: 2.10.0
  - files_reminders: 1.2.0
  - files_retention: 1.18.0
  - files_sharing: 1.21.0
  - files_trashbin: 1.19.0
  - google_synchronization: 3.0.0
  - groupfolders: 17.0.6
  - impersonate: 1.16.0
  - integration_github: 3.1.1
  - integration_google: 3.1.0
  - integration_mastodon: 2.0.6
  - logreader: 2.14.0
  - lookup_server_connector: 1.17.0
  - memories: 7.4.1
  - metadata: 0.21.0
  - music: 2.0.1
  - nextcloud_announcements: 1.18.0
  - nextcloudpi: 0.0.2
  - notifications: 2.17.0
  - notify_push: 0.7.0
  - oauth2: 1.17.0
  - password_policy: 1.19.0
  - picker: 1.0.11
  - previewgenerator: 99.99.99
  - privacy: 1.13.0
  - provisioning_api: 1.19.0
  - quota_warning: 1.20.0
  - recognize: 7.1.0
  - recommendations: 2.1.0
  - related_resources: 1.4.0
  - serverinfo: 1.19.0
  - settings: 1.12.0
  - sharebymail: 1.19.0
  - snappymail: 2.38.2
  - social: 0.8.0
  - socialsharing_email: 3.2.0
  - support: 1.12.0
  - survey_client: 1.17.0
  - suspicious_login: 7.0.0
  - tables: 0.8.1
  - theming: 2.4.0
  - twofactor_backupcodes: 1.18.0
  - updatenotification: 1.19.1
  - uppush: 1.5.1
  - user_status: 1.9.0
  - viewer: 2.3.0
  - weather_status: 1.9.0
  - workflowengine: 2.11.0
Disabled:
  - encryption: 2.17.0
  - epubviewer: 1.6.7 (installed 1.6.7)
  - files_versions: 1.22.0 (installed 1.22.0)
  - firstrunwizard: 2.18.0 (installed 2.15.0)
  - news: 25.0.2 (installed 25.0.2)
  - photos: 2.5.0 (installed 2.5.0)
  - systemtags: 1.19.0 (installed 1.19.0)
  - text: 3.10.1 (installed 3.10.1)
  - twofactor_totp: 11.0.0-dev
  - user_ldap: 1.20.0
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```json

```

### Additional info

I recall having this problem last November when I started using NC-27, there was something that was suggested and worked beautifully until this last minor update. I did try creating config/`mimetypealiases.json` with `text/html` mapped to 'file' and ran occ to update, but this had no effect.

I also thought I had the Text app added before, but when it is added, mobile doesn't even get a download link, they are given the text of the html sourcecode.

![Image](https://github.com/user-attachments/assets/9c110fe5-2b03-4792-ab9f-27bc70c202bb)

**Comment by teledyn** [comment]
There does not appear to be a solution for this; it has been suggested there is a problem with the `Content-Disposition` header, but I've no idea where this would be set. Even if someone could point me to an ugly hack where _all_ shared links to `text/html` simply download unmolested, I could _use_ Nextcloud again.
![Image](https://github.com/user-attachments/assets/a3823031-6d3a-428d-b469-6fbdf2787f60)

**Comment by teledyn** [comment]
Should anyone land here with the same issue, so far the only work-around is to use OwnCloud's Infinite Scale instead, or use a commercial service like GDrive or PCloud. 😒

The link generated by my nc is https://nas.teledyn.com/s/JtggMFbi5tPYBqT/download/Xmas%20Market.html which curl tells me has the following headers:
```
content-disposition: attachment; filename*=UTF-8''Xmas%20Market.html; filename="Xmas%20Market.html"
content-transfer-encoding: binary
content-type: text/plain;charset=UTF-8
```
which, again, downloads fine on desktop (so don't bother) and worked for a year on mobile, but is now transformed on download into a text rendering by Android Firefox, regardless the version of either Android or Firefox, and may do so on other platforms as I discovered this when my recipients reported the error, some likely on iPhones, and almost certainly none using Firefox.

OwnCloud has the same issue where I would _expect_ clicking the link in chat to evoke the download directly, prompting (as it did before) to open iReal as the handler, not the browser. OC still shunts to the browser to show the landing page partial preview, but in OC the link https://oc.teledyn.com/s/bpHJvfVFbUTXTJP redirects to `https://oc.teledyn.com/files/link/public/bpHJvfVFbUTXTJP?openWithDefaultApp=true&fileId=bpHJvfVFbUTXTJP&tiles-size=1&files-public-link-view-mode=resource-table&items-per-page=100` which offers the working download link `https://oc.teledyn.com/remote.php/dav/public-files/bpHJvfVFbUTXTJP/Xmas%20Market.html` that downloads immediately and unmodified and carries only these headers:
```
strict-transport-security: max-age=15768000; includeSubDomains
cache-control: no-cache
content-security-policy: frame-ancestors 'self'
content-type: text/html; charset=utf-8
date: Fri, 06 Dec 2024 17:57:32 GMT
last-modified: Fri, 06 Dec 2024 17:51:40 GMT
samesite: Strict
vary: Origin
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-request-id: nextcloudpi/ODJmjAI8K6-000186
x-web-version: 5.0.9
```

**Comment by teledyn** [comment]
For comparison, here are the headers from the equivalent link final download link as given by ownCloud; the only likely differences between them is ownCloud does not specify content-disposition, and also _correctly_ sets the mimetype to text/html
```
strict-transport-security: max-age=15768000; includeSubDomains
accept-ranges: bytes
content-length: 21608
content-security-policy: default-src 'none';
content-type: text/html
date: Sun, 08 Dec 2024 21:09:16 GMT
etag: "169851ee363e6f13879a4270c71a080e"
last-modified: Tue, 03 Dec 2024 04:39:17 GMT
oc-checksum: SHA1:8440374951c77c0607e3d2aa976f7c698591dcf6
oc-etag: "169851ee363e6f13879a4270c71a080e"
oc-fileid: 217043f7-9ebf-4ae5-ae27-8c8edcf54fe9$93913737-49ba-4496-99d8-5f4cf5eb320d!46ea0324-5f28-4b29-a48d-f3fcd3636f84
vary: Origin
x-content-type-options: nosniff
x-download-options: noopen
x-frame-options: SAMEORIGIN
x-permitted-cross-domain-policies: none
x-request-id: nextcloudpi/yrJVfs1WjA-000192
x-robots-tag: none
x-xss-protection: 1; mode=block
server: Apache
```

**Comment by teledyn** [comment]
An the Nextcloud puzzle is SOLVED! 🎉

Trouble is, I don't know how it was fixed. mimetypemapping.json to remove html to text/plain, so I cut it from the dist file too, ran occ mimetype-updates for js/db and reloaded apache/fpm, even cleared oc-cache! But it remained text/plain

So I just stepped through OCC fix commands:
maint:repair :theme:update files:cleanup :scan-app-data — the last is still running, so while waiting I tested, and curl now confirms content-type is html and mobile downloads and opens iReal!
